h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java

```diff
@@ -0,0 +1,336 @@
+      int origGamColStart = gamNumStart;
+      System.arraycopy(centerBeta,0, originalBeta, 0, gamColStart);   // copy everything before gamCols
+      for (int colInd=0; colInd < numGamCols; colInd++) {
+        double[] tempCbeta = new double[ztranspose[colInd].length];
+        System.arraycopy(centerBeta, gamColStart, tempCbeta, 0, tempCbeta.length);
+        double[] tempBeta = ArrayUtils.multVecArr(tempCbeta, ztranspose[colInd]);
+        System.arraycopy(tempBeta, 0, originalBeta, origGamColStart, tempBeta.length);
+        gamColStart += tempCbeta.length;
+        origGamColStart += tempBeta.length;
+      }
+      originalBeta[betaSize-1]=centerBeta[centerBeta.length-1];
+    } else
+      System.arraycopy(centerBeta, 0, originalBeta, 0, betaSize); // no change needed, just copy over
+
+    return originalBeta;
+  }
+
+  public static int copyGLMCoeffNames2GAMCoeffNames(GAMModel model, GLMModel glm, DataInfo dinfo) {
+    if (model._centerGAM) {
+      int numGamCols = model._gamColNames.length;
+      String[] glmColNames = glm._output.coefficientNames();
+      int lastGLMCoeffIndex = glmColNames.length-1;
+      int lastGAMCoeffIndex = dinfo.fullN();
+      int gamNumColStart = colIndexFromColNames(glmColNames, model._gamColNamesCenter[0][0]);
+      int gamLengthCopied = gamNumColStart;
+      System.arraycopy(glmColNames, 0, model._output._coefficient_names, 0, gamLengthCopied); // copy coeff names before gam columns
+      for (int gamColInd = 0; gamColInd < numGamCols; gamColInd++) {
+        System.arraycopy(model._gamColNames[gamColInd], 0, model._output._coefficient_names, gamLengthCopied,
+                model._gamColNames[gamColInd].length);
+        gamLengthCopied += model._gamColNames[gamColInd].length;
+      }
+      model._output._coefficient_names[lastGAMCoeffIndex] = new String(glmColNames[lastGLMCoeffIndex]);
+      return gamNumColStart;
+    } else
+      System.arraycopy(glm._output.coefficientNames(), 0, model._output._coefficient_names, 0,
+              dinfo.fullN()+1);
+    return 0;
+  }
+  public static void addGAM2Train(GAMParameters parms, Frame orig, double[][][] zTranspose,
+                                  double[][][] penalty_mat, String[][] gamColnames, String[][] gamColnamesCenter,
+                                  boolean modelBuilding, boolean centerGAM, Key<Frame>[] gamFrameKeys,
+                                  Key<Frame>[] gamFrameKeysCenter, double[][][] binvD, int[] numKnotsMat,
+                                  double[][] knotsMat, boolean saveZpenalty, double[][][] penalty_mat_noCenter) {
+    int numGamFrame = parms._gam_X.length;
+    boolean nullKnots = knotsMat == null;
+    RecursiveAction[] generateGamColumn = new RecursiveAction[numGamFrame];
+    for (int index = 0; index < numGamFrame; index++) {
+      final Frame predictVec = new Frame(new String[]{parms._gam_X[index]}, new Vec[]{orig.vec(parms._gam_X[index])});  // extract the vector to work on
```

Comment:
it seems that most of these are only used inside of the compute() function why not move it there and extract the RecursiveAction creation into separate method

Reply:
I used the recursiveAction to work on converting one column predictor into gam basis function values for each row in parallel.  Basically, the same set of operations are performed for each column predictor.  Hence, they are all there.  I did break down the different operations into functions to make them shorter.